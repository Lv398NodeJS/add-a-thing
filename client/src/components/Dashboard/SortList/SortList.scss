.btn {
  &.btn-no-outline {
    &:focus,
    &:active {
      outline: none !important;
      box-shadow: none !important;
    }
  }
}

.dropdown-toggle {
  &.no-arrow {
    &::after {
      display: none;
    }
  }
}

.dropdown-item {
  &.dont-highlight {
    &:active {
      background-color: #f8f9fa;
    }
  }
}

img.icon {
  height: 1.5em;
}

.dropdown-pull-up {
  position: absolute;
  top: 1rem;
  right: 0.8rem;
}
