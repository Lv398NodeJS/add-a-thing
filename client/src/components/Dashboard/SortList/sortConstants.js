export const sortDirections = {
  ASC: 'ASC',
  DESC: 'DESC',
  NONE: 'NONE',
};
