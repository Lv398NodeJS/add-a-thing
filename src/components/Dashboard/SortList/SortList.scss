.btn{
    &.btn-no-outline{
        &:focus,
        &:active {
            outline: none !important;
            box-shadow: none !important;
        }
    }

    &-custom-light{
        &:active,
        &.active {
            background-color: #d8d9da !important;
            border-color: #d8d9da !important;
        }
    }
}

.dropdown-item{
    &.dont-highlight{
        &:active {
            background-color: #f8f9fa;
        }
    }
}

img.icon{
    height: 1.5em;
}


.dropdown-pull-up{
    position: absolute;
    top: 1rem;
    right: 1.5rem;
}
.dropdown-pull-up{
    position: absolute;
    top: 1rem;
    right: 1.5rem;

    .dropdown-menu{
        top: -1.5rem !important;
        left: 0.75rem !important;
    }
}
