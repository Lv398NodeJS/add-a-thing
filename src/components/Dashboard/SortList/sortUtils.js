export const sortComparer = (field, direction = 'ASC') => (a, b) => {
  if (direction === 'NONE') {
    return 0;
  }

  let compareA = String(a[field]).toLowerCase();
  let compareB = String(b[field]).toLowerCase();

  if (field === 'priority') {
    compareA = String(['low', 'medium', 'high'].indexOf(compareA));
    compareB = String(['low', 'medium', 'high'].indexOf(compareB));
  }

  const directionMultiplier = direction === 'ASC' ? 1 : -1;
  return compareA.localeCompare(compareB) * directionMultiplier;
};

export const storage = {
  get: (key) => {
    let data = localStorage.getItem('SortList_Data') || '{}';
    data = JSON.parse(data);
    return data[key];
  },

  set: (key, value) => {
    const sortListDataRaw = localStorage.getItem('SortList_Data') || '{}';
    const sortListData = JSON.parse(sortListDataRaw);
    sortListData[key] = value;
    localStorage.setItem('SortList_Data', JSON.stringify(sortListData));
  },
};
