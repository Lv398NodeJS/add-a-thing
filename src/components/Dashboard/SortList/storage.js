const storage = {
  get: (key) => {
    let data = localStorage.getItem('SortList_Data') || '{}';
    data = JSON.parse(data);
    return data[key];
  },

  set: (key, value) => {
    let data = localStorage.getItem('SortList_Data') || '{}';
    data = JSON.parse(data);
    data[key] = value;
    data = JSON.stringify(data);
    localStorage.setItem('SortList_Data', data);
  },
};

export default storage;
