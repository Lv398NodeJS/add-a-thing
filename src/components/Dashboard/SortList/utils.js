export const sort = (array, field, direction = 'ASC') => {
  const copiedArray = [...array]; // For immutability

  copiedArray.sort((a, b) => {
    if (direction === 'NONE') {
      return 0;
    }

    let compareA = String(a[field]).toLowerCase();
    let compareB = String(b[field]).toLowerCase();

    if (field === 'priority') {
      compareA = ['Low', 'Medium', 'High'].indexOf(compareA);
      compareB = ['Low', 'Medium', 'High'].indexOf(compareB);
    }

    return compareA.localeCompare(compareB);
  });

  if (direction === 'DESC') {
    copiedArray.reverse();
  }

  return copiedArray;
};

export const storage = {
  get: (key) => {
    let data = localStorage.getItem('SortList_Data') || '{}';
    data = JSON.parse(data);
    return data[key];
  },

  set: (key, value) => {
    let data = localStorage.getItem('SortList_Data') || '{}';
    data = JSON.parse(data);
    data[key] = value;
    data = JSON.stringify(data);
    localStorage.setItem('SortList_Data', data);
  },
};
