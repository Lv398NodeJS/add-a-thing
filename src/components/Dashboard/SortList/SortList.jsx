import React, { Component } from 'react';
import {
  Dropdown,
  ButtonGroup,
  Button,
} from 'react-bootstrap';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as sortListActions from '../../../actions/sortListActions';
import './SortList.scss';
import { ReactComponent as SortIcon } from './sort.svg';
import { ReactComponent as SortIconAsc } from './sort-asc.svg';
import { ReactComponent as SortIconDesc } from './sort-desc.svg';

class SortList extends Component {
  constructor(props) {
    super(props);
    const { storageKey, sortListActions: { loadSort } } = this.props;

    loadSort(storageKey);
  }

  setSort = (newField, newDirection) => {
    const { storageKey, allSortData, sortListActions: { setSort } } = this.props;
    const { field, direction } = allSortData[this.key] || {};

    if (field === newField && direction === newDirection) {
      setSort(storageKey, 'NONE', 'NONE');
    } else {
      setSort(storageKey, newField, newDirection);
    }
  };

  render() {
    const { color, allSortData, storageKey } = this.props;
    const { field, direction } = allSortData[storageKey] || {};

    return (
      <Dropdown
        alignRight
        className="dropdown-pull-up"
      >
        <Dropdown.Toggle
          className="btn-no-outline no-arrow"
          size="sm"
          variant="none"
        >
          <SortIcon fill={color} />
        </Dropdown.Toggle>

        <Dropdown.Menu>
          <Dropdown.Item
            className="btn-group btn-group-vertical dont-highlight p-0"
            key="name"
          >
            <ButtonGroup
              size="sm"
            >
              <Button
                variant=""
                className="rounded-0 text-left disabled btn-no-outline"
              >
                Name
              </Button>
              <Button
                variant="light"
                active={field === 'name' && direction === 'ASC'}
                className="rounded-0 btn-no-outline flex-grow-0"
                onClick={() => this.setSort('name', 'ASC')}
              >
                <SortIconAsc />
              </Button>
              <Button
                variant="light"
                active={field === 'name' && direction === 'DESC'}
                className="rounded-0 btn-no-outline flex-grow-0"
                onClick={() => this.setSort('name', 'DESC')}
              >
                <SortIconDesc />
              </Button>
            </ButtonGroup>
          </Dropdown.Item>
          <Dropdown.Item
            className="btn-group btn-group-vertical dont-highlight p-0"
            key="priorityForSorting"
          >
            <ButtonGroup
              size="sm"
            >
              <Button
                variant=""
                className="rounded-0 text-left disabled btn-no-outline"
              >
                Priority
              </Button>
              <Button
                variant="light"
                active={field === 'priority' && direction === 'ASC'}
                className="rounded-0 btn-no-outline flex-grow-0"
                onClick={() => this.setSort('priority', 'ASC')}
              >
                <SortIconAsc />
              </Button>
              <Button
                variant="light"
                active={field === 'priority' && direction === 'DESC'}
                className="rounded-0 btn-no-outline flex-grow-0"
                onClick={() => this.setSort('priority', 'DESC')}
              >
                <SortIconDesc />
              </Button>
            </ButtonGroup>
          </Dropdown.Item>
        </Dropdown.Menu>
      </Dropdown>
    );
  }
}

const mapStateToProps = ({ sortListReducer }) => ({
  allSortData: sortListReducer,
});
const mapDispatchToProps = dispatch => ({
  sortListActions: bindActionCreators(sortListActions, dispatch),
});

export { SortList as SortListComponent };

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SortList);
