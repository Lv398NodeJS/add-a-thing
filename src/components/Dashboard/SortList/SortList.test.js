import React from 'react';
import { shallow, render } from 'enzyme';
import SortList from './SortList';

const TestChild = props => (
  <span>
    {JSON.stringify(props, null, 2)}
  </span>
);

const sortList = shallow(
  <SortList
    storageKey="test"
    sortIconColor="#ffc000"
    fields={[
      {
        key: 'priorityForSorting',
        text: 'Priority',
      },
    ]}
  >
    <TestChild priorityForSorting="3">c</TestChild>
    <TestChild priorityForSorting="1">a</TestChild>
    <TestChild priorityForSorting="2">b</TestChild>
  </SortList>,
);

describe('<SortList />', () => {
  it('should render and match snapshot', () => {
    const renderTest = render(
      <SortList
        storageKey="test"
        sortIconColor="#ffc000"
        fields={[
          {
            key: 'priorityForSorting',
            text: 'Priority',
          },
        ]}
      >
        <TestChild priorityForSorting="3">c</TestChild>
        <TestChild priorityForSorting="1">a</TestChild>
        <TestChild priorityForSorting="2">b</TestChild>
      </SortList>,
    );
    expect(renderTest).toMatchSnapshot();
  });

  it('should render dropdown', () => {
    const dropdown = sortList.find('Dropdown');
    expect(dropdown.length).toBe(1);
  });

  it('should render children', () => {
    const dropdown = sortList.find('TestChild');
    expect(dropdown.length).toBe(3);
  });
});
