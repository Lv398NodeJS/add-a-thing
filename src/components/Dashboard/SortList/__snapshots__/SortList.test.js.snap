// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SortList /> should render and match snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "dropdown-pull-up dropdown",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "aria-expanded": "false",
          "aria-haspopup": "true",
          "class": "btn-no-outline no-arrow dropdown-toggle btn btn-none btn-sm",
          "type": "button",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "fill": "#ffc000",
            },
            "children": Array [
              Object {
                "data": "sort.svg",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "svg",
            "namespace": "http://www.w3.org/2000/svg",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "fill": undefined,
            },
            "x-attribsPrefix": Object {
              "fill": undefined,
            },
          },
        ],
        "name": "button",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "aria-expanded": undefined,
          "aria-haspopup": undefined,
          "class": undefined,
          "type": undefined,
        },
        "x-attribsPrefix": Object {
          "aria-expanded": undefined,
          "aria-haspopup": undefined,
          "class": undefined,
          "type": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": Object {
      "attribs": Object {},
      "children": Array [
        Object {
          "attribs": Object {},
          "children": Array [
            Object {
              "data": "{
  \\"priorityForSorting\\": \\"3\\",
  \\"children\\": \\"c\\"
}",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "data": "{
  \\"priorityForSorting\\": \\"1\\",
  \\"children\\": \\"a\\"
}",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {},
              "children": Array [
                Object {
                  "data": "{
  \\"priorityForSorting\\": \\"2\\",
  \\"children\\": \\"b\\"
}",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "span",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        Object {
          "attribs": Object {},
          "children": Array [
            Object {
              "data": "{
  \\"priorityForSorting\\": \\"1\\",
  \\"children\\": \\"a\\"
}",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "data": "{
  \\"priorityForSorting\\": \\"2\\",
  \\"children\\": \\"b\\"
}",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "data": "{
  \\"priorityForSorting\\": \\"3\\",
  \\"children\\": \\"c\\"
}",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        Object {
          "attribs": Object {},
          "children": Array [
            Object {
              "data": "{
  \\"priorityForSorting\\": \\"2\\",
  \\"children\\": \\"b\\"
}",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "data": "{
  \\"priorityForSorting\\": \\"1\\",
  \\"children\\": \\"a\\"
}",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {},
              "children": Array [
                Object {
                  "data": "{
  \\"priorityForSorting\\": \\"3\\",
  \\"children\\": \\"c\\"
}",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "span",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "div",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": [Circular],
      "root": Object {
        "attribs": Object {},
        "children": Array [
          [Circular],
          [Circular],
        ],
        "name": "root",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": null,
        "prev": null,
        "type": "root",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      "type": "tag",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
        Object {
          "attribs": Object {},
          "children": Array [
            Object {
              "attribs": Object {},
              "children": Array [
                Object {
                  "data": "{
  \\"priorityForSorting\\": \\"3\\",
  \\"children\\": \\"c\\"
}",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "span",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "data": "{
  \\"priorityForSorting\\": \\"1\\",
  \\"children\\": \\"a\\"
}",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "span",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": Object {
                  "attribs": Object {},
                  "children": Array [
                    Object {
                      "data": "{
  \\"priorityForSorting\\": \\"2\\",
  \\"children\\": \\"b\\"
}",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "span",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Object {
              "attribs": Object {},
              "children": Array [
                Object {
                  "data": "{
  \\"priorityForSorting\\": \\"1\\",
  \\"children\\": \\"a\\"
}",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "span",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "data": "{
  \\"priorityForSorting\\": \\"2\\",
  \\"children\\": \\"b\\"
}",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "span",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "data": "{
  \\"priorityForSorting\\": \\"3\\",
  \\"children\\": \\"c\\"
}",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "span",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Object {
              "attribs": Object {},
              "children": Array [
                Object {
                  "data": "{
  \\"priorityForSorting\\": \\"2\\",
  \\"children\\": \\"b\\"
}",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "span",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "data": "{
  \\"priorityForSorting\\": \\"1\\",
  \\"children\\": \\"a\\"
}",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "span",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": Object {
                  "attribs": Object {},
                  "children": Array [
                    Object {
                      "data": "{
  \\"priorityForSorting\\": \\"3\\",
  \\"children\\": \\"c\\"
}",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "span",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": null,
          "prev": [Circular],
          "root": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "1": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {},
        "children": Array [
          Object {
            "data": "{
  \\"priorityForSorting\\": \\"3\\",
  \\"children\\": \\"c\\"
}",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {},
          "children": Array [
            Object {
              "data": "{
  \\"priorityForSorting\\": \\"1\\",
  \\"children\\": \\"a\\"
}",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "data": "{
  \\"priorityForSorting\\": \\"2\\",
  \\"children\\": \\"b\\"
}",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      Object {
        "attribs": Object {},
        "children": Array [
          Object {
            "data": "{
  \\"priorityForSorting\\": \\"1\\",
  \\"children\\": \\"a\\"
}",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {},
          "children": Array [
            Object {
              "data": "{
  \\"priorityForSorting\\": \\"2\\",
  \\"children\\": \\"b\\"
}",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {},
          "children": Array [
            Object {
              "data": "{
  \\"priorityForSorting\\": \\"3\\",
  \\"children\\": \\"c\\"
}",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      Object {
        "attribs": Object {},
        "children": Array [
          Object {
            "data": "{
  \\"priorityForSorting\\": \\"2\\",
  \\"children\\": \\"b\\"
}",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {},
          "children": Array [
            Object {
              "data": "{
  \\"priorityForSorting\\": \\"1\\",
  \\"children\\": \\"a\\"
}",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "data": "{
  \\"priorityForSorting\\": \\"3\\",
  \\"children\\": \\"c\\"
}",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": Object {
      "attribs": Object {
        "class": "dropdown-pull-up dropdown",
      },
      "children": Array [
        Object {
          "attribs": Object {
            "aria-expanded": "false",
            "aria-haspopup": "true",
            "class": "btn-no-outline no-arrow dropdown-toggle btn btn-none btn-sm",
            "type": "button",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "fill": "#ffc000",
              },
              "children": Array [
                Object {
                  "data": "sort.svg",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "svg",
              "namespace": "http://www.w3.org/2000/svg",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "fill": undefined,
              },
              "x-attribsPrefix": Object {
                "fill": undefined,
              },
            },
          ],
          "name": "button",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "aria-expanded": undefined,
            "aria-haspopup": undefined,
            "class": undefined,
            "type": undefined,
          },
          "x-attribsPrefix": Object {
            "aria-expanded": undefined,
            "aria-haspopup": undefined,
            "class": undefined,
            "type": undefined,
          },
        },
      ],
      "name": "div",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": [Circular],
      "parent": null,
      "prev": null,
      "root": Object {
        "attribs": Object {},
        "children": Array [
          [Circular],
          [Circular],
        ],
        "name": "root",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": null,
        "prev": null,
        "type": "root",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      "type": "tag",
      "x-attribsNamespace": Object {
        "class": undefined,
      },
      "x-attribsPrefix": Object {
        "class": undefined,
      },
    },
    "root": Object {
      "attribs": Object {},
      "children": Array [
        Object {
          "attribs": Object {
            "class": "dropdown-pull-up dropdown",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "aria-expanded": "false",
                "aria-haspopup": "true",
                "class": "btn-no-outline no-arrow dropdown-toggle btn btn-none btn-sm",
                "type": "button",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "fill": "#ffc000",
                  },
                  "children": Array [
                    Object {
                      "data": "sort.svg",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "svg",
                  "namespace": "http://www.w3.org/2000/svg",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "fill": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "fill": undefined,
                  },
                },
              ],
              "name": "button",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "aria-expanded": undefined,
                "aria-haspopup": undefined,
                "class": undefined,
                "type": undefined,
              },
              "x-attribsPrefix": Object {
                "aria-expanded": undefined,
                "aria-haspopup": undefined,
                "class": undefined,
                "type": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": null,
          "prev": null,
          "root": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 2,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
